### PR TITLE
Added Dask/pyramid suport. Split tiff/imagej/ome reader at metadata level instead.

### PR DESCRIPTION
**Pull request by folterj**
WIP, TODO: processing ome metadata

### CODE REVIEW

File: napari_tiff/_tests/test_tiff_reader.py [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
 import zipfile
 
-import numpy as np
 from napari_tiff import napari_get_reader
 from napari_tiff.napari_tiff_reader import (imagecodecs_reader,
-                                            imagej_reader,
                                             tifffile_reader,
                                             zip_reader)
-import pytest
-import tifffile
```
Comment (jni) [inline napari_tiff/_tests/test_tiff_reader.py:11]: Why did you remove the imagej_reader from tests?

File: napari_tiff/napari_tiff_reader.py [hunk 2]
```diff
@@ -299,7 +313,27 @@ def imagej_reader(tif):
...
-    return [(data, kwargs, 'image')]
+    return kwargs
+
+
+def get_ome_tiff_metadata(tif):
+    metadata = xml2dict(tif.ome_metadata)
+    if 'OME' in metadata:
+        metadata = metadata['OME']
+
+    # TODO: process ome metadata
```
Comment (jni) [inline napari_tiff/napari_tiff_reader.py:324]: It might be worth depending on @tlambert03's [ome-types](https://github.com/tlambert03/ome-types) for this.

File: napari_tiff/_tests/test_tiff_reader.py (hunk 1 above)
Comment (jni) [reply napari_tiff/_tests/test_tiff_reader.py:11]: Oh, I see, you're trying to just unify everything under one reader that automagically does the right thing. That's fine, but I think it's worth keeping and testing the other readers as independent functions. This makes it easier to understand when one particular functionality is broken.

File: napari_tiff/_tests/test_tiff_reader.py [hunk 3]
```diff
@@ -67,7 +66,6 @@ def test_reader(tmp_path, data_fixture, original_data):
 
 @pytest.mark.parametrize("reader, data_fixture, original_data", [
     (imagecodecs_reader, example_data_filepath, np.random.random((20, 20))),
-    (imagej_reader, example_data_tiff,  np.random.randint(0, 255, size=(20, 20)).astype(np.uint8)),
```
Comment (GenevieveBuckley) [inline napari_tiff/_tests/test_tiff_reader.py:70]: It's important that we don't lose functionality, so we'll need to keep this test.

File: pyproject.toml [hunk 4]
```diff
@@ -29,7 +29,8 @@ requires-python = '>=3.10'
 dependencies = [
    'imagecodecs',
    'numpy',
-   'tifffile>=2020.5.7',
+   'tifffile>=2023.9.26',
```
Comment (GenevieveBuckley) [inline pyproject.toml:32]: Is there a specific feature/requirement causing the bump in version number?

File: napari_tiff/_tests/test_tiff_reader.py (hunk 3 above)
Comment (GenevieveBuckley) [reply napari_tiff/_tests/test_tiff_reader.py:70]: ...ah, I think I see what's happening with this restructure now

**Comment by GenevieveBuckley** [comment]
Could this be a possible failure case? https://forum.image.sc/t/tifffile-opening-individual-ome-tiff-files-as-single-huge-array-even-when-isolated/77701

@jni do you happen to remember or have a link to those files, so we can check?

File: napari_tiff/_tests/test_tiff_reader.py (hunk 1 above)
Comment (GenevieveBuckley) [reply napari_tiff/_tests/test_tiff_reader.py:11]: This PR removes imagej_reader entirely, and instead has separate functions for finding imagej/ome/tiff metadata (instead of having separate functions for reading each of those types of tiff files).

File: napari_tiff/napari_tiff_reader.py [hunk 5]
```diff
@@ -72,14 +70,30 @@ def zip_reader(path: PathLike) -> List[LayerData]:
 
 
 def tifffile_reader(tif):
-    """Return napari LayerData from largest image series in TIFF file."""
+    """Return napari LayerData from image series in TIFF file."""
+    nlevels = len(tif.series[0])
+    if nlevels > 1:
+        data = [da.from_zarr(tif.aszarr(level=level)) for level in range(nlevels)]
+    else:
+        data = da.from_zarr(tif.aszarr())
```
Comment (GenevieveBuckley) [inline napari_tiff/napari_tiff_reader.py:78]: I'm not entirely sure how I feel about making both dask and zarr hard requirements for opening any tiff, even the non-pyramidal ones

File: napari_tiff/napari_tiff_reader.py [hunk 6]
```diff
@@ -299,7 +313,27 @@ def imagej_reader(tif):
...
+
+def get_ome_tiff_metadata(tif):
+    metadata = xml2dict(tif.ome_metadata)
+    if 'OME' in metadata:
+        metadata = metadata['OME']
+
+    # TODO: process ome metadata
+
+    kwargs = dict(
+        rgb=rgb,
```
Comment (GenevieveBuckley) [inline napari_tiff/napari_tiff_reader.py:327]: It looks like these variables are not defined in this function? (`rgb`, `channel_axis`, `name`, `scale`, `colormap`, `contrast_limits`, `blending`, `visible`, etc.)

File: napari_tiff/napari_tiff_reader.py [hunk 7]
```diff
@@ -299,7 +313,27 @@ def imagej_reader(tif):
...
         visible=visible,
     )
-    return [(data, kwargs, 'image')]
+    return kwargs
+
+
+def get_ome_tiff_metadata(tif):
+    metadata = xml2dict(tif.ome_metadata)
+    if 'OME' in metadata:
+        metadata = metadata['OME']
```
Comment (GenevieveBuckley) [inline napari_tiff/napari_tiff_reader.py:322]: The `metadata` variable doesn't seem to get used in this function after this?

File: napari_tiff/napari_tiff_reader.py [hunk 8]
```diff
@@ -72,14 +70,30 @@ def zip_reader(path: PathLike) -> List[LayerData]:
 
 
 def tifffile_reader(tif):
-    """Return napari LayerData from largest image series in TIFF file."""
+    """Return napari LayerData from image series in TIFF file."""
+    nlevels = len(tif.series[0])
```
Comment (GenevieveBuckley) [inline napari_tiff/napari_tiff_reader.py:74]: I'm still a little confused about the difference between tiff series and tiff pages. Is anyone else able to explain it?

File: napari_tiff/napari_tiff_reader.py [hunk 9]
```diff
@@ -72,14 +70,30 @@ def zip_reader(path: PathLike) -> List[LayerData]:
...
+    if nlevels > 1:
+        data = [da.from_zarr(tif.aszarr(level=level)) for level in range(nlevels)]
```
Comment (GenevieveBuckley) [inline napari_tiff/napari_tiff_reader.py:76]: This assumes that the tiff series are arranged in a pyramid format. Is that something that is absolutely guaranteed by the tiff specification? 
If not, we may need to check we actually have a multi-resolution file with all the resolution levels in the right order before passing it to napari.

File: napari_tiff/_tests/test_tiff_reader.py (hunk 1 above)
Comment (folterj) [reply napari_tiff/_tests/test_tiff_reader.py:11]: Ok it also saves imagej metadata - I didn't know that was supported. I'll put back the test.

File: napari_tiff/napari_tiff_reader.py (hunk 2 above)
Comment (folterj) [reply napari_tiff/napari_tiff_reader.py:324]: I understood this is mainly used for formatting/writing as opposed to reading. We have a dictionary at this point so not sure what including the ome-types package would add here?

File: napari_tiff/_tests/test_tiff_reader.py (hunk 3 above)
Comment (folterj) [reply napari_tiff/_tests/test_tiff_reader.py:70]: See above

File: pyproject.toml (hunk 4 above)
Comment (folterj) [reply pyproject.toml:32]: Yes, loading (and lazy writing) levels / zarr / dask

File: napari_tiff/napari_tiff_reader.py [hunk 10]
```diff
@@ -72,14 +70,30 @@ def zip_reader(path: PathLike) -> List[LayerData]:
...
+    else:
+        data = da.from_zarr(tif.aszarr())
+    if tif.is_ome:
+        kwargs = get_ome_tiff_metadata(tif)
+    elif tif.is_imagej:
+        kwargs = get_imagej_metadata(tif)
+    else:
+        kwargs = get_tiff_metadata(tif)
```
[the remark this answers is not in the document]
Comment (folterj) [reply napari_tiff/napari_tiff_reader.py:84]: Yes that's also possible. This is exactly what I'm doing in our image conversion pipeline e.g.:

https://github.com/FrancisCrickInstitute/OmeSliCC/blob/f6e2aecd5713bbbff7e10e04eebe5ede4bace1cc/OmeSliCC/TiffSource.py#L113

File: napari_tiff/napari_tiff_reader.py (hunk 5 above)
Comment (folterj) [reply napari_tiff/napari_tiff_reader.py:78]: This is what allows lazy loading & pyramid sizes.
Although this works regardless, I'll split it up and put back asarray() for single series.

File: napari_tiff/napari_tiff_reader.py (hunk 8 above)
Comment (folterj) [reply napari_tiff/napari_tiff_reader.py:74]: Although @cgohlke is the expert as creator of tifffile, as I understand it pages is the internal structure of tiff, and the series/levels provide higher level access (which point to the underlying pages).

File: napari_tiff/napari_tiff_reader.py (hunk 9 above)
Comment (folterj) [reply napari_tiff/napari_tiff_reader.py:76]: I've not seen this written otherwise in any WSI file format. I imagine it's spec.

File: napari_tiff/napari_tiff_reader.py (hunk 2 above)
Comment (tlambert03) [reply napari_tiff/napari_tiff_reader.py:325]: ome-types does both, and inasmuch as you're only going to be pulling a couple things like channel names, colors, and scales from the metadata, I agree that it's probably not necessary to bring on an additional dependency just to grab a few things.

you will, however, likely eventually run into malformed xml that doesn't strictly meet the ome data model.  And in that case you'll just want to make sure that you essentially assume very little about the structure of the dict you get back (and fail gracefully).  ome-types tests itself against a ton of xml examples and might do slightly better in that regard, and the object you get back will have a guaranteed structure (and fixes a handful of commonly seen errors in various ome xml implementations, like micro-manager).  

I'd say go with what tifffile provides for now, and if you get any bug reports with poorly-mined metadata, then try out ome-types on those files and see if it makes things any more robust

**Comment by jni** [comment]
> Could this be a possible failure case? [forum.image.sc/t/tifffile-opening-individual-ome-tiff-files-as-single-huge-array-even-when-isolated/77701](https://forum.image.sc/t/tifffile-opening-individual-ome-tiff-files-as-single-huge-array-even-when-isolated/77701)

Yes it could.

> @jni do you happen to remember or have a link to those files, so we can check?

Here's the data, for manual testing, not broad distribution:

https://www.dropbox.com/scl/fo/rqopeqntdsgzep4r10qbs/h?rlkey=k931yg05j4sbnqlh60z343k74&dl=0

**Comment by jni** [comment]
Having said this, since @folterj is using dask/zarr and not imread, it probably would work but would return empty slices where the data is missing. And that is arguably the correct behaviour here.

File: napari_tiff/napari_tiff_reader.py (hunk 9 above)
Comment (Czaki) [reply napari_tiff/napari_tiff_reader.py:76]: Doctring from series `Series of pages with compatible shape and data type.` So if the length of series is different from 1 then different shapes are present. 

However, I'm not sure if it always should be read as pyramidal. As I know that tiff allows storing thumbnail for example. But I do not have such data available.

File: napari_tiff/_tests/test_tiff_reader.py [hunk 11]
```diff
@@ -18,9 +17,21 @@ def example_data_filepath(tmp_path, original_data):
...
-    tifffile.imwrite(example_data_filepath, original_data, imagej=True)
-    return tifffile.TiffFile(example_data_filepath)
+    filepath = str(tmp_path / "myfile.tif")
+    tifffile.imwrite(filepath, original_data)
+    return tifffile.TiffFile(filepath)
+
+
+def example_data_imagej(tmp_path, original_data):
+    filepath = str(tmp_path / "myfile.tif")
+    tifffile.imwrite(filepath, original_data, imagej=True)
```
[the remark this answers is not in the document]
Comment (folterj) [reply napari_tiff/_tests/test_tiff_reader.py:27]: This is a good point. The original test code read plain tif as imagej, so trying to expand with some simple imagej (and ome-tiff for ome-tiff reading) metadata that tifffile supports.
I'm happy to leave if/what metadata the imagej part of the code would be tested against, as I'm focussing on the ome-tiff reading.
In general a concise, non-generated dataset is needed for more thorough testing.

[the remark this answers is not in the document]
Comment (Czaki) [reply napari_tiff/_tests/test_tiff_reader.py:27]: Same remark is to ome function. It is also without providing any metadata.

File: napari_tiff/napari_tiff_reader.py (hunk 9 above)
Comment (folterj) [reply napari_tiff/napari_tiff_reader.py:76]: @Czaki this is testing the length / number of levels inside the first serie. Would a thumbnail be the same or a separate series? What would the best way be to test for pyramid - ome metadata?

Comment (Czaki) [reply napari_tiff/napari_tiff_reader.py:76]: I do not know. I work with writing my own tiff parser a long time ago and cannot find the data that I used then. This is rather point that it will be nice to check this (for example using some public datasets). I will also try to digg it later.

**Comment by GenevieveBuckley** [comment]
Superseded by https://github.com/napari/napari-tiff/pull/24
